### PR TITLE
Workflow for building container image

### PR DESCRIPTION
**Pull request by gvegayon**
This pull request introduces a new GitHub Actions workflow for building Docker images and refactors the Dockerfiles to separate dependencies from the main application. The key changes include the creation of a workflow file, modifications to the `Containerfile`, and the addition of a new `Containerfile.dependencies`.

### Workflow and Dockerfile changes:

* **New GitHub Actions workflow**:
  * Added `.github/workflows/containers.yaml` to automate the building and pushing of Docker images. This workflow includes steps to checkout code, retrieve commit messages, determine branch names, and handle Docker caching and login.

* **Refactoring Dockerfiles**:
  * Modified `Containerfile` to use a pre-built dependencies image from the container registry, allowing for faster builds and better separation of concerns.
  * Added a new `Containerfile.dependencies` to define the base image with all necessary dependencies, including Python, R, and CMake installations.

> [!IMPORTANT]
> This depends on the `cfa-cdcgov` runners, which are currently not set for this repository. If you don't want to depend on the cdcgov runner (for now), we can change this to use ghcr.io instead.

### CODE REVIEW

File: Containerfile.dependencies [hunk 1]
```diff
@@ -0,0 +1,9 @@
+FROM python:3.12
```
Comment (damonbayer) [inline Containerfile.dependencies:1]: Let's update to 3.13 while we're making changes?

Comment (damonbayer) [inline Containerfile.dependencies:1]:
```suggestion
FROM python:3.13
```
Let's update to 3.13 while we're making changes?

**Comment by jkislin** [comment]
@gvegayon  @damonbayer you've been setup with two instances of `cfa-cdcgov` for this repository. 
Please let me know if you experience issues

> [!IMPORTANT]
> Please do not invite non-cdc users to be admins of this repository, or to allow non-cdc users to push directly to branches that trigger CI/CD.

**Comment by gvegayon** [comment]
> @gvegayon @damonbayer you've been setup with two instances of `cfa-cdcgov` for this repository. Please let me know if you experience issues
> 
> Important
> 
> Please do not invite non-cdc users to be admins of this repository, or to allow non-cdc users to push directly to branches that trigger CI/CD.

@damonbayer and @dylanhmorris, aren't there explicit rules in the repo for that? 

attn @zsusswein and @natemcintosh

**Comment by gvegayon** [comment]
The thing is runnning! But there's an issue with pyarrow. Will investigate.

**Comment by gvegayon** [comment]
Hey @damonbayer! It turns out that python 3.13 breaks the container (b/c of some stuff with pyarrow). Downgrading to Python 3.12 works.

File: Containerfile.dependencies [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+RUN apt-get update
+RUN apt-get install -y r-base
+RUN apt-get install -y cmake
+RUN pip install --root-user-action=ignore -U pip
+RUN pip install --root-user-action=ignore git+https://github.com/cdcgov/pyrenew
```
Comment (gvegayon) [inline Containerfile.dependencies:7]: Also, why not installing a fixed version of pyrenew? Is there a release yet, @damonbayer?

Comment (damonbayer) [reply Containerfile.dependencies:7]: Just released [v0.1.1](https://github.com/CDCgov/PyRenew/releases/tag/v0.1.1), which we can target.

**Comment by natemcintosh** [comment]
> Hey @damonbayer! It turns out that python 3.13 breaks the container (b/c of some stuff with pyarrow). Downgrading to Python 3.12 works.

I also had some issues with pyarrow when upgrading to 3.13. I was able to fix them by updating from pyarrow 17 to 18.

File: Containerfile.dependencies (hunk 2 above)
Comment (gvegayon) [reply Containerfile.dependencies]: Just added it as an argument as well.

**Comment by damonbayer** [comment]
@gvegayon I update pyarrow. Please try Python 3.13 again, if you don't mind.

**Comment by gvegayon** [comment]
> @gvegayon I update pyarrow. Please try Python 3.13 again, if you don't mind.

Running again!

**Comment by natemcintosh** [comment]
If Damon merged the change to 3.13 into main, did that change also get merged into this branch?

**Comment by gvegayon** [comment]
> If Damon merged the change to 3.13 into main, did that change also get merged into this branch?. Otherwise this branch will still be trying to use pyarrow 17, no?

I don't think so. GHA workflows triggered by PRs run on a merge (see [here](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request)). I can try actively merging main into the Branch, but I think it will be the same thing.

**Comment by gvegayon** [comment]
@natemcintosh and @damonbayer, it's passing now! I'm still learning how PR-triggered workflows work, but at least is working!

**Review by damonbayer — APPROVED** [review]

**Comment by damonbayer** [comment]
Thanks @gvegayon!

**Pull request merged**